text/3414-deprecate-spooky-dropck-at-a-distance.md

```diff
@@ -0,0 +1,362 @@
+        let s = String::from("temporary");
+        let p = PrintOnDrop(&s);
+        x = (to_pd(p), String::new());
+    }
+}
+```
+
+And yet, this compiles:
+
+```rust
+use core::marker::PhantomData;
+
+struct PrintOnDrop<'s>(&'s str);
+impl<'s> Drop for PrintOnDrop<'s> {
+    fn drop(&mut self) {
+        println!("{}", self.0);
+    }
+}
+
+fn to_pd<T>(_: T) -> PhantomData<T> {
+    PhantomData
+}
+
+pub fn foo() {
+    let mut x;
+    {
+        let s = String::from("temporary");
+        let p = PrintOnDrop(&s);
+        x = (to_pd(p), ());
+    }
+}
+```
+
+`PhantomData`'s dropck behaviour is only checked if the type (in this case, the
+tuple) marks `needs_drop`, which is confusing. Unrelated tuple elements
+shouldn't affect `PhantomData` behaviour, but the above example shows that they
+do. This RFC makes it so they don't.
+
+Likewise, `[T; 0]` produces the same effect: it is not `needs_drop`, but adds
+outlive requirements, thus also exhibiting "spooky-dropck-at-a-distance".
+
+Simply defining every non-`needs_drop` type as being pure w.r.t. drop would,
+however, break `#[may_dangle]`, so we need to adapt it.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Changes to dropck (stable)
+
+Types like `fn(T)`, `ManuallyDrop<T>`, `PhantomData<T>`, `[T; 0]` and `&'a T`,
+that don't need to be dropped, place no liveness requirements on `'a` or `T`
+when they go out of scope - even if `T` would otherwise introduce liveness
+requirements.
+
+In other words, this compiles:
+
+```rust
+fn main() {
+    let mut x;
+    {
+        x = &String::new();
+        // String implicitly dropped here
+    }
+    // x implicitly dropped here
+}
+```
+
+But a type which does need to be dropped, introduces liveness requirements:
+
+```rust
+struct PrintOnDrop<'s>(&'s str);
+impl<'s> Drop for PrintOnDrop<'s> {
+    fn drop(&mut self) {
+        println!("{}", self.0);
+    }
+}
+
+fn main() {
+    let mut x;
+    {
+        x = PrintOnDrop(&*String::new());
+        // String implicitly dropped here
+    }
+    // x implicitly dropped here
+    // ERROR: temporary may not live long enough
+}
+```
+
+Unless it's in one of the above types:
+
+```rust
+struct PrintOnDrop<'s>(&'s str);
+impl<'s> Drop for PrintOnDrop<'s> {
+    fn drop(&mut self) {
+        println!("{}", self.0);
+    }
+}
+
+fn main() {
+    let mut x;
+    {
+        x = &PrintOnDrop(&*String::new());
+        // PrintOnDrop implicitly dropped here
+        // String implicitly dropped here
+    }
+    // x implicitly dropped here
+}
+```
+
+As a special case, some built-in types, like `Vec`, `Box`, `Rc`, `Arc`,
+`HashMap`, among others, despite needing to be dropped, do not place liveness
+requirements unless `T` demands liveness requirements. This is okay:
+
+```rust
+fn main() {
+    let mut x = vec![];
+    {
+        x.push(&String::new());
+        // String implicitly dropped here
+    }
+    // x implicitly dropped here
+}
+```
+
+But this isn't:
```

Comment:
I generally find that this style is confusing for the reader, as they must now diff the two examples to understood what is being contrasted. Here and elsewhere, can you highlight specifically what you are comparing?

```suggestion
But this example errors, because the vector now contains a `PrintOnDrop<'_>`, which implements `Drop`:
```

In this case, I'd probably add comments into the code too.